### PR TITLE
Split up repl tests to make them faster

### PR DESCRIPTION
**Pull request by cocreature**
This PR splits up the tests into the tests for TLS and Auth and the
tests for the actual functionality.

The func tests use the repl as a library which allows them to be
significantly faster:

1. We only need to start the service process once.
2. We only need to initialize the package db once.

There is at least one other point that I did not address for now and
that is only loading the packages into the repl service once. While
loading them multiple times is a noop, it still has a performance
implication.

Sadly, this has turned out much more messy than I thought it would be
due to various issues with haskeline/repline/tasty/computers. The
details are in a comment in DA.Test.Repl.FuncTests.

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by cocreature** [comment]
Ahrg, this is failing on Windows with `{handle: <file descriptor: 5>}: hDuplicateTo: illegal operation (handles are incompatible)`. Apparently I’m not allowed to `dup2` between file handles and pipes. Not sure what to do about this :sob:

**Comment by cocreature** [comment]
I guess I can try redirecting to a file since the statements don’t depend on each other (atm, I was originally hoping to make that possible at some point).

File: compiler/damlc/tests/src/DA/Test/Repl/FuncTests.hs [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+main = do
+    scriptDar <- locateRunfiles (mainWorkspace </> "daml-script" </> "daml" </> "daml-script.dar")
+    testDar <- locateRunfiles (mainWorkspace </> "compiler" </> "damlc" </> "tests" </> "repl-test.dar")
+    replDir <- locateRunfiles (mainWorkspace </> "compiler/repl-service/server")
+    forM_ [stdin, stdout, stderr] $ \h -> hSetBuffering h LineBuffering
+    let replJar = replDir </> "repl-service.jar"
+    -- Turning the repl client into a resource is a bit annoying since it requires
+    -- pulling apart functions like `withGrpcClient`. Therefore we just
+    -- allocate the resources before handing over to tasty and accept that
+    -- -l will spin up sandbox and the repl client.
```
Comment (aherrmann-da) [inline compiler/damlc/tests/src/DA/Test/Repl/FuncTests.hs:54]: What's `-l`?

File: compiler/damlc/tests/src/DA/Test/Repl/FuncTests.hs [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+
+matchServiceOutput :: String -> Step
+matchServiceOutput s = MatchOutput (makeRegex s) s ReplService
+
+readPrompt :: Handle -> Expectation
+readPrompt h = do
+    res <- replicateM 6 (hGetChar h)
+    res `shouldBe` "daml> "
+
+-- | Based on the function in `main-tester` of the same name which is sadly not exposed.
```
Comment (aherrmann-da) [inline compiler/damlc/tests/src/DA/Test/Repl/FuncTests.hs:279]:
```suggestion
-- | Based on the function in @main-tester@ of the same name which is sadly not exposed.
```
haddock markup

**Comment by cocreature** [comment]
> AFAIK that's what `silently` does https://hackage.haskell.org/package/silently-1.2.5.1/docs/System-IO-Silently.html

Yeah I was hoping to be able to be clever here and still keep the interactivity by using a pipe but apparently I cannot on Windows.

**Comment by cocreature** [comment]
It kind of works now on Windows but there are two flaky issues:
1. Sometimes the last line of output from the service (not the client) is missing so it locks up forever. I have no idea why that is. I disabled buffering completely without any effect. I wonder if I need to flush somewhere in the logback config but I couldn’t really figure out how.
2. Sometimes the `hGetLine` on the client output is failing with
```
  1) repl func tests module imports
       uncaught exception: IOException of type EOF
       /tmp/extra-file-2153013822790-14-4: hGetLine: end of file
```

Both of those issues seem flaky and I understand neither of them.

File: compiler/damlc/tests/src/DA/Test/Repl/FuncTests.hs (hunk 1 above)
Comment (cocreature) [reply compiler/damlc/tests/src/DA/Test/Repl/FuncTests.hs]: An impressively bad typo of `it` :joy:

**Comment by cocreature** [comment]
Finally got it working! The issue was actually completely unrelated to this PR and instead related to the already existing issue around the echo diagnostics. I didn’t completely figure out what was going wrong but something around diagnostics and document versions caused a race condition where the newly added import sometimes did produce a type error based on the old file. I’ve switched over to `setBufferModified` which seems nicer than writing a file anyway, fixes the echo problems and fixes the issues I was seeing here.

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs [hunk 3]
```diff
@@ -249,15 +245,11 @@ runRepl opts mainDar replClient ideState = do
...
         let newImports = imp : imports
         -- TODO[AH] Factor out the module render and typecheck step.
-        liftIO $ writeFileUTF8 (fromNormalizedFilePath $ lineFilePath lineNumber)
-            (renderModule dflags newImports lineNumber [] (WildPat noExt)
+        liftIO $ setBufferModified ideState (lineFilePath lineNumber)
+            $ Just $ T.pack (renderModule dflags newImports lineNumber [] (WildPat noExt)
                 (noLoc $ HsApp noExt
                     (noLoc $ HsVar noExt $ noLoc $ mkRdrUnqual $ mkVarOcc "return")
                     (noLoc $ ExplicitTuple noExt [] Boxed)))
-        -- Useful for debugging, probably best to put it behind a --debug flag
```
Comment (cocreature) [inline compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs:257]: These are out of date now and trying to ensure code in a comment seems a bit of a lost cause so I’ve removed them for now. Happy to add them back under a debugging flag separately.

**Pull request merged**